pact/fixed-quote-policy-royalty.pact

```diff
@@ -0,0 +1,192 @@
+    )
+
+  (defschema quote-schema
+    id:string
+    spec:object{quote-spec})
+
+  (deftable quotes:{quote-schema})
+
+  (defun get-policy:object{policy-schema} (token:object{token-info})
+    (read policies (at 'id token))
+  )
+
+  (defun enforce-ledger:bool ()
+     (enforce-guard (marmalade.ledger.ledger-guard))
+   )
+
+  (defun enforce-mint:bool
+    ( token:object{token-info}
+      account:string
+      amount:decimal
+    )
+    (enforce-ledger)
+    (bind (get-policy token)
+      { 'mint-guard:=mint-guard:guard
+      , 'max-supply:=max-supply:decimal
+      }
+      (enforce-guard (at 'mint-guard (get-policy token)))
+      (enforce (<= (+ amount (at 'supply token)) max-supply) "Exceeds max supply")
+  ))
+
+  (defun enforce-burn:bool
+    ( token:object{token-info}
+      account:string
+      amount:decimal
+    )
+    (enforce-ledger)
+    (enforce false "Burn prohibited")
+  )
+
+  (defun enforce-init:bool
+    ( token:object{token-info}
+    )
+    (enforce-ledger)
+    (let* ( (fungible:module{fungible-v2} (read-msg 'fungible ))
+            (creator:string (read-msg 'creator ))
+            (creator-guard:guard (read-keyset 'creator-guard ))
+            (mint-guard:guard (read-keyset 'mint-guard ))
+            (max-supply:decimal (read-decimal 'max-supply ))
+            (min-amount:decimal (read-decimal 'min-amount ))
+            (royalty-rate:decimal (read-decimal 'royalty-rate ))
+            (creator-details:object (fungible::details creator ))
+            )
+      (fungible::precision)
+      (enforce (=
+        (at 'guard creator-details) creator-guard)
+        "Creator guard does not match")
+      (enforce (and
+        (>= royalty-rate 0.0) (<= royalty-rate 1.0))
+        "royalty rate is not valid")
+      (insert policies (at 'id token)
+        { 'fungible: fungible
+        , 'creator: creator
+        , 'creator-guard: creator-guard
+        , 'mint-guard: mint-guard
+        , 'max-supply: max-supply
+        , 'min-amount: min-amount
+        , 'royalty-rate: royalty-rate }))
+    true
+  )
+
+  (defun enforce-offer:bool
+    ( token:object{token-info}
+      seller:string
+      amount:decimal
+      sale-id:string
+    )
+    @doc "Capture quote spec for SALE of TOKEN from message"
+    (enforce-ledger)
+    (enforce-sale-pact sale-id)
+    (bind (get-policy token)
+      { 'fungible := fungible:module{fungible-v2}
+      }
+    (let* ( (spec:object{quote-spec} (read-msg QUOTE))
+            (price:decimal (at 'price spec))
+            (recipient:string (at 'recipient spec))
+            (recipient-guard:guard (at 'recipient-guard spec))
+            (recipient-details:object (fungible::details recipient)) )
+      (fungible::enforce-unit price)
+      (enforce (< 0.0 price) "Offer amount must be positive")
+      (enforce (=
+        (at 'guard recipient-details) recipient-guard)
+        "Recipient guard does not match")
+      (insert quotes sale-id { 'id: (at 'id token), 'spec: spec }))
```

Comment:
We need to enforce that the royalty rate doesn't run into precision here. Ie a royalty rate of 1% with a fungible precision of 4 and a price 0.0001 would result in an invalid royalty (so the seller must increase the quote price).

Reply:
Enforced `(* sale-price royalty-rate)`. Also added `(enforce-unit royalty-rate)` in `enforce-init`